Author: John-Mark Bell <jmb@pexip.com>
Date  : Mon Feb 25 17:49:43 2013 +0000

    Handle time going backwards by using the monotonic clock

    THIRD-PARTY: Apache 2.0

--- a/lib/puppet/scheduler/timer.rb
+++ b/lib/puppet/scheduler/timer.rb
@@ -1,4 +1,12 @@
 module Puppet::Scheduler
+  CLOCK_GETTIME = 228
+  CLOCK_MONOTONIC_RAW = 4
+
   class Timer
+    def initialize(*args)
+      @epoch = Time.now
+      @base = self.monotime
+    end
+
     def wait_for(seconds)
       if seconds > 0
@@ -7,7 +15,14 @@ module Puppet::Scheduler
     end
 
     def now
-      Time.now
+      @epoch + (self.monotime - @base)
+    end
+
+   def monotime
+      buf = [0x1111111111111111, 0x1111111111111111].pack "Q2"
+      syscall CLOCK_GETTIME, CLOCK_MONOTONIC_RAW, buf
+      ts = buf.unpack "Q2"
+      return ts[0] + ts[1]/1000000000.0
     end
   end
 end
